#summary This page is under construction.

= Introduction =

This page will describe how create draft genome assemblies from raw Fastq reads. 

= Prerequisites = 
 * [http://aluru-sun.ece.iastate.edu/doku.php?id=reptile Reptile] - error-correction software 
 * [http://i.cs.hku.hk/~alse/hkubrg/projects/idba/ IDBA] - Iterative DeBruijn graph de novo assembler
 * [http://bio-bwa.sourceforge.net/ BWA] - short-read aligner using Burrows-Wheeler transform
 * [http://www.physics.rutgers.edu/~anirvans/SOPRA/ SOPRA] - statistical-optimization scaffolding
 * repair - pairs reads named using Illumina format sequence identifiers. Found in ngopt repository under trunk/tools/pair_reads
 * rmctgs - removes small contigs from Fasta file. Found in ngopt repository under trunk/tools/seqtools. 
 * scaf_sopra_bwa.sh - wrapper script for scaffolding with BWA and SOPRA. Found in ngopt repository under trunk/tools/assembly_line.
 * assemblePE.sh - runs IDBA-BWA-SOPRA pipeline. Found in ngopt repository under trunk/tools/assembly_line

= Assembling Paired Reads in FastQ = 
* For the rest of this tutorial, we will be assembling
== Before we get started... ==
You will need to shuffle your paired reads if they are not already shuffled. To do so, run *`repair`* like so:
{{{
repair --paired --shuf -s .fastq pe_shuf reads.fastq
}}} 
where `pe_shuf` is the output basename for our shuffled reads. This will produce a file named `pe_shuf.fastq` in the current working directory.  The `-s` option or `repair`  can be used to output to a different directory.
== Reptile Error Correction == 
#Convert Fastq to fasta+qual files. In the directory containing `pe_shuf.fastq`, run:
{{{
fastaq-converter.pl ./ ./  2 
}}}