#summary A simple example of preprocessing of multiplexed Illumina.

= Introduction =
This page will describe how to preprocess custom-barcode multiplexed Illumina data. 
= Prerequisites = 
You will need the following software in your executable path.
 * [http://genome.gsc.riken.jp/osc/english/dataresource/ Tagdust] - eliminate artifactual sequence from NGS data
 * [http://code.google.com/p/ngopt/source/browse/trunk/tools/splitbc/splitBC?r=70 splitBC] - a script to split barcoded reads. 
If your reads are paired, you will also need the following program:
 * [http://code.google.com/p/ngopt/source/browse/trunk/tools/pair_reads/repair.cpp?r=85 repair] - a program for pairing and shuffling Illumina reads

You will need the following sequence files:
 * [http://code.google.com/p/ngopt/source/browse/trunk/adapter.fasta?r=85 adapter.fasta] - Illumina adapter sequences.
 * a tab-delimited file with your barcode sequences. This file should look something like this:
 {{{
 BC1    ATGCTA
 BC2    TTGACT
 BC3    GTAGGT
 }}}
 * a single file containing all of your reads. They do not need to be shuffled if you have paired reads
= Preprocess Multiplexed Data =
== Remove Artifactual DNA ==
Often times during library preparations, DNA fragments consisting only of adapter sequence are inserted and sequenced. To get rid of reads containing these sequences, we will use the program `tagdust`. 
{{{
tagdust -o reads.clean.fastq -a reads.artifact.fastq adapters.fasta reads.fastq
}}}
where `reads.fastq` are our Illumina reads. Reads without adapter sequence will be output to `reads.clean.fastq`
== Split Barcoded Reads == 
To split our barcoded data set, we will use the `splitBC` from ngopt. 
{{{
splitBC reads.clean.fastq --bcfile barcodes.txt --prefix ./ --suffix .fastq --bol 
}}}
