#summary A simple example for correcting errors in Illumina reads

= This page is under construction =
= Introduction =

This page will describe how to use NGOpt tools with Reptile to correct errors in Illumina reads.

= Prerequisites =
You will need the following packages/programs. They will need to be in your executable path:
 * [http://aluru-sun.ece.iastate.edu/doku.php?id=reptile Reptile] - error-correction software 

== Step 1.  Convert Fastq to Fasta+Qual Files.== 
In the directory containing `pe_shuf.fastq`, run:
  {{{
  <ngopt_trunk>/tools/quality/reptile_fq-converter ./ ./  2 
  }}}
   This will create *two* files, `pe_shuf.fa` and `pe_shuf.q`
  Note: All files ending in .fastq in the current directory will be processed.
== Step 2. Create Reptile Config-File ==
Copy the Reptile config-file template from the Reptile directory and change as needed.
  {{{
  cp <path_to_reptile>/src/config~ ./pe_shuf.rec.conf
  vim pe_shuf.rec.conf
  }}}
 Edit `pe_shuf.rec.conf` to specify input and output files. 
  * Modify line beginning with "`InFaFile`"  to contain `pe_shuf.fa`
  * Modify line beginning with "`IQFile`" to contain `pe_shuf.q`  
  * Modify line beginning with "`OErrFile`" to contain the desired name of the output file. 
   * This output file will not be corrected reads. We are going to run another command on this file to get our corrected reads. Let's call the file `reptile-output` 
 If you are more educated in the Reptile Error Correction process, feel free to modify any of other run configurations. 
== Step 3. Run Reptile Error Correction ==
Run `reptile` with the config file from previous step.
  {{{
  reptile pe_shuf.rec.conf
  }}}
== Step 4. Merge Corrections ==
{{{
  reptile_merger pe_shuf.fa reptile-output pe_shuf.rec.fasta
}}}

 Voila! Our corrected reads are now located in `pe_shuf.rec.fasta`