#summary A simple example for correcting errors in Illumina reads

= Introduction =

This page will describe how to use NGOpt tools with Reptile to correct errors in Illumina reads.

= Step 1. Reformat Fastq Files = 
 # Convert Fastq to fasta+qual files. In the directory containing `pe_shuf.fastq`, run:
  {{{
  <path_to_reptile>/scripts/fastq-converter.pl ./ ./  2 
  }}}
   This will create *two* files, `pe_shuf.fa` and `pe_shuf.q`
  Note: All files ending in .fastq in the current directory will be processed.
 # Copy and edit Reptile config-file template.
  {{{
  cp <path_to_reptile>/src/config~ ./pe_shuf.rec.conf
  vim pe_shuf.rec.conf
  }}}
  Edit `pe_shuf.rec.conf` to specify input and output files. 
   * Modify line beginning with "`InFaFile`"  to contain `pe_shuf.fa`
   * Modify line beginning with "`IQFile`" to contain `pe_shuf.q`  
   * Modify line beginning with "`OErrFile`" to contain the desired name of the output file. 
    * This output file will not be corrected reads. We are going to run another command on this file to get our corrected reads. Let's call the file `reptile-output` 
  If you are more educated in the Reptile Error Correction process, feel free to modify any of other run configurations. 
 # Run Reptile error correction
  {{{
  reptile pe_shuf.rec.conf
  reptile_merger pe_shuf.fa reptile-output pe_shuf.rec.fasta
  }}}
 Voila! Our corrected reads are now located in `pe_shuf.rec.fasta`