\documentclass{bioinfo}
\copyrightyear{2011}
\pubyear{2011}

\begin{document}
\firstpage{1}

\title[a5]{a5}
\author[ChangeMe \textit{et~al}]{ChangeMe\,$^{1}$\footnote{to whom correspondence should be addressed}, Andrew Tritt\,$^{1}$ Jonathan A. Eisen\,$^{1,2,3}$ and Marc T. Facciotti\,$^{1,4}$}
\address{$^{1}$Genome Center, $^{2}$ Dept. of Evolution and Ecology, $^{3}$ Medical Microbiology and Immunology, 
$^{4}$ Biomedical Engineering, University of California-Davis, Davis, CA 95616.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}

\section{Summary:}
blahblah
\section{Availability:}
GPL source code and a usage tutorial is at \href{http://ngopt.googlecode.com}{http://ngopt.googlecode.com}

\section{Contact:} \href{blahblah}{blahblah}
\end{abstract}

\section{Introduction}
Despite many efforts to solve the problem of genome assembly, the task still remains non-trivial. 

Rapid growth of high throughput DNA sequence techonologies has made sequence 
data readily available..... Analyzing this data is difficult, in part because 
assembly algorithms have many parameters that are not easily optimized.... 

We introduce a new de novo genome assembly pipeline that uses previously 
published assembly and other short read tools. 

\begin{methods}
\section{Methods}
The genome assembly problem can be broken down into two main steps: 1) building
contigs and 2) scaffolding contigs. 

We separate building contigs into two steps. The first step in building contigs
is read error correction. For this step we use error correction tools from the
SGA software package/pipeline. After reads are corrected for sequencing errors,
contigs are built using the assembler IDBA. (Do we discuss why we use IDBA?)  

After contigs are built, paired/mated libraries can be used to scaffold contigs 
together. For this stage, our pipeline uses the scaffolder SSPACE. A key 
parameter for scaffolding is the insert size of libraries. Using the read
mapping software package BWA, insert size estimations have been automated. Another
key parameter in 


\end{methods}

\subsection{Automated misassembly quality control}

After assembling genomes, an important step is quality control and validation
of "assembly connections". The final step in our pipeline is automated 
misassembly detection. For this step, we use FISH to identify erroneously
connected points in the genome. The FISH algorithm was orignally developed
to identify collinear segments of homology between two genomes. Here, we use
FISH to identify collinear segments of paired read connections. First, reads 
are mapped onto scaffolds using BWA and connected segments are identified using 
FISH. Scaffolds are then broken at the boundaries of these segments and the 
resulting contigs/scaffolds are rescaffolded using SSPACE as described above.


\subsection{Automated parameter selection}




\begin{table}[!t]
\processtable{ Metrics on assemblies of Haloferax volcanii from four pipelines.
\label{Tab:01}}
{\begin{tabular}{l|cccc}\toprule
Metric & volcS & volcV & volcA5 & volcA5-noQC\\\midrule
Scaffold count & 1000 & 10000 & 50  & 6 \\
Miscalled bases & 1000 & 2e6 & 100 & 100 \\
Uncalled bases & 15000 & 60000 & 10000 & 7500 \\
Extra bases & 5.0\% & 2.5\% & 8.0\% & 4.0\% \\
Missing bases & 7.5\% & 13.0\% & 7.5\% & 7.0\% \\
Extra segments & 500 & 262 & 45 & 1 \\
Missing segments & 117 & 1144 & 192 & abc \\
DCJ Distance & 1100 & 10250 & 56 & 8 \\
Intact CDS & 93.2\% & 89.4\% & 97.0\% & 97.3\% \\
\botrule \\
\end{tabular}}{}
\end{table}



\section{Discussion}



\begin{figure}[t]
%\includegraphics[width=3.5in]{missingextra_merge.eps}
\vspace{-1cm}
\caption{\textbf{Top:} Density of extra and missing segments in the assemblies of \textit{Haloferax volcanii} DS2.
Reference genome coordinates are given on the x-axis, and red vertical bars delineate the boundaries of the
five circular replicons in the reference genome. \textbf{Bottom:} Size distribution of missing and extra segments in each assembly.  The size of a missing segment is given on the x-axis, and the count of missing segments at that size on the y-axis.}\label{fig:01}
\end{figure}

\begin{figure}[t]
%\includegraphics[width=3.5in]{mauve_basecall_bias.eps}
\vspace{-1cm}
\caption{Biased errors in the base calling of each assembly. Errors are not uniformly random in any of the three assemblies. See Supplementary Material for more details.}\label{fig:02}
\end{figure}



\section*{Acknowledgements}
This work was supported by National Science Foundation award ER 0949453.

\bibliographystyle{natbib}
\bibliography{ngopt-appnote}

\end{document}
